torch/csrc/autograd/function.cpp

```diff
@@ -35,96 +35,33 @@ AnomalyMetadata* Function::metadata() noexcept {
- *
- * The solution here is to use a custom deleter with each
- * std::shared_ptr<Function>. The custom deleter keeps track of how many
- * nested deleters it is in. When this number exceeds the maximum allowed
- * depth, the Function* to be deleted are accumulated in a per-thread
- * delete queue and handled by one of the deleters.
- *
- * Note that these custom deleters are NOT necessary for deleting PyFunction.
- * This is because a THPFunction Python object owns a PyFunction that is in a
- * computation graph. When Python objects get recursively destroyed, they
- * are also queued into a delete list. This happens very early for them
- * (at 50 deleters): https://github.com/python/cpython/blob/f320be77ffb73e3b9e7fc98c37b8df3975d84b40/Include/object.h#L1024-L1063
- * so we don't need to worry about them.
- */
-
-thread_local std::deque<Function*> deleteFunctionQueue;
-thread_local size_t deleteFunctionRecursionDepth = 0;
-
-/*
- * If this number is set too high, a deep computation graph can still
- * stack overflow. The procedure for setting this number was to
- * 1) find the smallest value that would not guard against stack overflows
- *    on various machines
- * 2) Take the minimum of all such values and subtract some leeway because
- *    the memory of these stack frames will probably grow as time passes.
- * Testing on a few machines machines, the magic numbers were:
- * - Mac OSX (Macbook Pro 15) : ~60000
- * - A beefy Ubuntu 16.04 box : ~15000
- * - Windows AWS instance (g3.4xlarge): variable. My two attempts at different
- *   times have gotten the following numbers: ~8300, 3669
- */
-#ifdef _WIN32
-constexpr size_t kDeleteFunctionMaxRecursionDepth = 3000;
-#else
-constexpr size_t kDeleteFunctionMaxRecursionDepth = 10000;
-#endif
-
-struct RecursionDepthCounter {
- public:
-  explicit RecursionDepthCounter() {
-    ++deleteFunctionRecursionDepth;
-  }
-  ~RecursionDepthCounter() {
-    --deleteFunctionRecursionDepth;
-  }
-
-  size_t value() {
-    return deleteFunctionRecursionDepth;
+static void gatherFunctions(Function* func,
+                            std::vector<std::shared_ptr<Function>>& queue) {
+  for (auto edge : func->next_edges()) {
+    if (edge.function.get()) {
+      queue.emplace_back(std::move(edge.function));
+    }
   }
-};
+}
 
-/*
- * Note that the custom deleter deletes in BFS style. Without using
- * the custom deleter, the computation graph is deleted in a DFS style.
- * The BFS deletion is valid (and safe) because if a shared_ptr<Function>
- * 's reference count hits 0, nothing else will access it.
- */
 void deleteFunction(Function* function) {
-  RecursionDepthCounter recursion_depth;
-
-  if (recursion_depth.value() > kDeleteFunctionMaxRecursionDepth) {
-    deleteFunctionQueue.push_back(function);
-    return;
-  }
-
+  // To avoid stack overflow on large computational graphs,
+  // we need to track reference decrementing and freeing
+  // on the heap.
+  std::vector<std::shared_ptr<Function>> queue;
```

Comment:
Err... this isn't a queue anymore, right? Because you replaced it with a vector, it's now a stack, and you're doing it DFS.

Reply:
Weirdly enough, we *did* talk about this in the original code review: https://github.com/pytorch/pytorch/pull/6873#pullrequestreview-115413722 but I guess we forgot to stop using a `deque` when we switched back to stack discipline.

Reply:
There is some commentary in the original code that indicates that it was deliberately doing a BFS traversal, which I assumed was the impetus for using a deque.  std::vector doesn't have a pop_front().